pybamm/solvers/idaklu_jax.py

```diff
@@ -0,0 +1,956 @@
+import pybamm
+import numpy as np
+import logging
+import warnings
+import numbers
+
+from typing import Callable
+from typing import Union
+from typing import List
+
+from functools import lru_cache
+
+import importlib.util
+import importlib
+
+idaklu_spec = importlib.util.find_spec("pybamm.solvers.idaklu")
+if idaklu_spec is not None:
+    try:
+        idaklu = importlib.util.module_from_spec(idaklu_spec)
+        if idaklu_spec.loader:
+            idaklu_spec.loader.exec_module(idaklu)
+    except ImportError:  # pragma: no cover
+        idaklu_spec = None
+
+if pybamm.have_jax():
+    import jax
+    from jax import lax
+    from jax import numpy as jnp
+    from jax.interpreters import ad
+    from jax.interpreters import mlir
+    from jax.interpreters import batching
+    from jax.interpreters.mlir import custom_call
+    from jax.lib import xla_client
+    from jax.tree_util import tree_flatten
+
+
+class IDAKLUJax:
+    """
+    JAX wrapper for IDAKLU solver
+
+    Objects of this class should be created via an IDAKLUSolver object.
+
+    Parameters
+    ----------
+    solver : :class:`pybamm.IDAKLUSolver`
+        The IDAKLU solver object to be wrapped
+    """
+
+    def __init__(
+        self,
+        solver,
+    ):
+        if not pybamm.have_jax():
+            raise ModuleNotFoundError(
+                "Jax or jaxlib is not installed, please see https://docs.pybamm.org/en/latest/source/user_guide/installation/gnu-linux-mac.html#optional-jaxsolver"
+            )  # pragma: no cover
+        if not pybamm.have_idaklu():
+            raise ModuleNotFoundError(
+                "IDAKLU is not installed, please see https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html"
+            )  # pragma: no cover
+        self.jaxpr = None  # JAX expression
+        self.idaklu_jax_obj = None  # IDAKLU-JAX object
+        self.solver = solver  # Originating IDAKLU Solver object
+
+    def get_jaxpr(self):
+        """Returns a JAX expression representing the IDAKLU-wrapped solver object"""
+        if self.jaxpr is None:
+            raise pybamm.SolverError("jaxify() must be called before get_jaxpr()")
+        return self.jaxpr
+
+    def get_var(
+        self,
+        f: Callable,
+        varname: str,
+    ):
+        """Helper function to extract a single variable from the jaxified expression
+
+        Returns a JAX expression
+
+        Parameters
+        ----------
+        f : function
```

Comment:
Yes, this has been refactored a few times and at some point was a static method that confirmed to the typical jax functional form of `x(y(f))(a, b, c)`. However, since the jaxified expression is now bound to the idaklu-jax instance this requirement can be dropped and the `f` in the arguments list can be removed and replaced with the jaxpr. As such I've now removed the `f` such that this function now effectively replaces `f` in user calls when filtering the results by variable name(s). I have amended the docs (specifically the jupyter notebook) and tests to reflect this.

Reply:
I'm going to have to row-back on this one. If we use only `jaxpr` for the expression then we cannot pass higher-order functions that the user may have constructed (a prominent example being `jax.jit(f)` into this function); it was designed to be a helper function to isolate variables from any expression (not just the base jaxpr).

Reply:
On further testing I've decided to refactor this again to accept different call signatures. Now, you *can* call it with only the variable name of interest (in which case it will use jaxpr); but you can also provide a JAX expression (useful if you want to perform some other function, e.g. `jax.jit(f)`) before isolating the variable, or you can provide an `np.ndarray`, which allows you to extract derivates from functions like `jacfwd` more easily. I think this provides the best of both worlds, and have added an example use-case in the jupyter notebook for the latter case, specifically:
```python
d = jax.jacfwd(f, argnums=1)(t_eval, inputs)  # returns a dict of np.ndarrays
data = jax_solver.get_var(d["Current function [A]"], "Voltage [V]")  # Isolates derviate of V wrt current fcn
```